### PR TITLE
[media-library] Fix a crash when calling `getAssetInfoAsync` on slow motion videos on iOS

### PR DESCRIPTION
**Pull request by jarvisluong**
# Why and How

Described in: https://github.com/expo/expo/issues/8801

# Test Plan

Choose to see details of a slow motion video on iOS Bare app, the app should not crash

### CODE REVIEW

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 1]
```diff
@@ -422,9 +422,31 @@ - (NSDictionary *)constantsToExport
       [[PHImageManager defaultManager] requestAVAssetForVideo:asset
                                                       options:options
                                                 resultHandler:^(AVAsset * _Nullable asset, AVAudioMix * _Nullable audioMix, NSDictionary * _Nullable info) {
-        AVURLAsset *urlAsset = (AVURLAsset *)asset;
-        result[@"localUri"] = [[urlAsset URL] absoluteString];
-        resolve(result);
+        // Slow motion videos are returned as an AVComposition instance
+        if ([asset isKindOfClass:[AVComposition class]]) {
+            NSString *directory = [self.fileSystem.cachesDirectory stringByAppendingPathComponent:@"MediaLibrary"];
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:427]: Please extract this `MediaLibrary` string to a const variable.

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 2]
```diff
@@ -422,9 +422,31 @@ - (NSDictionary *)constantsToExport
...
+            [self.fileSystem ensureDirExistsWithPath:directory];
+            NSString *videoOutoutFileName = [NSString stringWithFormat:@"slowMoVideo-%d.mov",arc4random() % 1000];
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:429]:
```suggestion
            NSString *videoOutputFileName = [NSString stringWithFormat:@"slowMoVideo-%d.mov",arc4random() % 1000];
```

And is it possible to get the original file name?

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 3]
```diff
@@ -422,9 +422,31 @@ - (NSDictionary *)constantsToExport
...
+            NSString *videoFileOutputPath = [directory stringByAppendingPathComponent:videoOutoutFileName];
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:430]:
```suggestion
            NSString *videoFileOutputPath = [directory stringByAppendingPathComponent:videoOutputFileName];
```

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 4]
```diff
@@ -422,9 +422,31 @@ - (NSDictionary *)constantsToExport
...
+            NSURL *videoFileOutputURL = [NSURL fileURLWithPath:videoFileOutputPath];
+            
+            AVAssetExportSession *exporter = [[AVAssetExportSession alloc] initWithAsset:asset presetName:AVAssetExportPresetHighestQuality];
+            exporter.outputURL = videoFileOutputURL;
+            exporter.outputFileType = AVFileTypeQuickTimeMovie;
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:435]: Here, we are sure that the output file type is `QuickTimeMovie`, aren't we? I'm asking cause I'm not sure if the `AVComposition` can store only slow-motion videos.

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 5]
```diff
@@ -422,9 +422,31 @@ - (NSDictionary *)constantsToExport
...
+            exporter.shouldOptimizeForNetworkUse = YES;
+                                
+            [exporter exportAsynchronouslyWithCompletionHandler:^{
+                if (exporter.status == AVAssetExportSessionStatusCompleted) {
+                    result[@"localUri"] = videoFileOutputURL.absoluteString;
+                    resolve(result);
+                }
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:442]: We should reject if the status is `AVAssetExportSessionStatusFailed` or `AVAssetExportSessionStatusCancelled `.

**Review by lukmccall — CHANGES_REQUESTED** [review]
Overall it looks good 👍

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m (hunk 4 above)
Comment (jarvisluong) [reply packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:435]: According to the definition on Apple’s documentation, AVComposition is meant for storing special “videos”, so yes I think exporting to QuickTimeMovie type is reasonable 🤔 https://developer.apple.com/documentation/avfoundation/avcomposition

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m (hunk 2 above)
Comment (jarvisluong) [reply packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:429]: I don’t see any helpful method to get the “file name” yet:

PHAsset class: https://developer.apple.com/documentation/photokit/phasset

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m (hunk 4 above)
Comment (lukmccall) [reply packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:435]: Special "videos" is a little bit ambiguous 😅 But exporting as `QuickTimeMovie` seems reasonable in this case ;)

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m [hunk 6]
```diff
@@ -24,6 +24,8 @@
 
 NSString *const EXMediaLibraryDidChangeEvent = @"mediaLibraryDidChange";
 
+NSString *const EXMediaLibraryCachesDirectory = @"MediaLibrary";
```
Comment (lukmccall) [inline packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:27]: I'm not sure if this variable name should be prefixed with `EX`. It won't break anything. However, I don't think we want to have a separate cache folder for each SDK version. cc @tsapeta

**Review by lukmccall — COMMENTED** [review]
Thanks for applying all requested changes 🎉

File: packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m (hunk 6 above)
Comment (jarvisluong) [reply packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:27]: @tsapeta could you provide some input?

Comment (tsapeta) [reply packages/expo-media-library/ios/EXMediaLibrary/EXMediaLibrary.m:27]: Sorry for slow reaction 😅 It's fine as is – FileSystem's cacheDirectory should already be versioned so I think we don't need to do this again here.
(For clarification, references prefixes by `EX` are being versioned by our tools when we add new SDKs to the project. In other words, `EXMediaLibrary` would be transformed to `ABI38_0_0EXMediaLibrary`, but `MediaLibrary` would remain unchanged.)

**Review by tsapeta — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]

**Comment by awinograd** [comment]
Thanks for your fixes to expo-media-library @jarvisluong. They've been working perfectly for me. 

Any chance we can publish a new version of the package, expo team? I'm currently manually applying this patch when building my app. On a general note, what's the normal cadence for publishing updates like this? Happy to wait until the normal time if there's a standard process.

Thanks all!

**Comment by tsapeta** [comment]
Yeah, sure, I've just published a new version — 8.3.0 🎉

We don't have any strict cadence for publishing updates other than just before SDK releases, however I'm trying to improve this. Before SDK38 we were publishing all packages at once and only during the release process, but now we're slowly pushing more module-centric approach forward to make our modules even more attractive than the alternative ones 😉 As expo modules often depend on others, we've had some concerns about publishing them more frequently, mostly due to compatibility reasons. However, thanks to some new tools that I've built recently, I hope that one day we will achieve our goal which is to publish them regularly (once per week?) by GitHub Actions — but until then, just let me know if you're waiting for the fix 😄 

Unfortunately only bare workflow projects will benefit from this because we cannot release the Expo Client to the stores as often as we would like 😞

**Comment by awinograd** [comment]
Thank you, @tsapeta! Much appreciated.

**Comment by mvjansen** [comment]
Hi @tsapeta, is there any way to get this bug fix working in an expo project (i.e. not a bare work flow project)?  I have installed the 8.3.0 version of expo-media-library and created a new fresh ios build, but my app is still crashing on slow mo videos

**Comment by jarvisluong** [comment]
> Unfortunately only bare workflow projects will benefit from this because we cannot release the Expo Client to the stores as often as we would like 😞 

Since this is a native change, you will need for a new expo client to include this fix

**Comment by mvjansen** [comment]
Hi @jarvisluong, I saw the comment for an updated expo client, but thought a build of a new .ipa file would incorporate the new native code and not have any dependence on the expo client when running as a standalone app.  Obviously I'm missing something?

**Comment by jarvisluong** [comment]
IMO the ipa build on Expo service should basically freeze the version of the dependencies to make sure the build is stable. I think it's best that you eject to the bare workflow and then be able to build the new native version with the fix. I believe the eject process is not as painful as it used to be (in fact I use bare workflow for all projects now)

**Pull request merged**